File: FlappyClone/api/validation.py

# NB: If these constants are updated, remember to update the JavaScript versions (in js/common/validation.js)!!
NUM_LEADERBOARD_ENTRIES = 10
MAX_NAME_LENGTH = 16
LEGAL_SYMBOLS = '-_'

# NB: If updating these functions, ensure that the JavaScript functions are also updated (in js/common/validation.js)!
def isValidNameChar(c):
    if (c >= 'a' and c <= 'z'):
        return True
    elif (c >= 'A' and c <= 'Z'):
        return True
    elif (c >= '1' and c <= '9'):
        return True
    elif (LEGAL_SYMBOLS.find(c) != -1):
        return True
    else:
        return False

# NB: If updating these functions, ensure that the JavaScript functions are also updated (in js/common/validation.js)!
def isValidName(name):
    reason = ''
    if (len(name) == 0):
        return (False, 'name is an empty string')
    elif (len(name) > MAX_NAME_LENGTH):
        return (False, 'name is too long (' + len(name) + ' characters, max is ' + MAX_NAME_LENGTH + ')')
    else:
        for c in name:
            if (not isValidNameChar(c)):
                return (False, 'name contains an invalid character (' + c + ')')
    return True

File: static/js/flappy/game/validation.js
// NB: If these constants are updated, remember to update the Django versions (in api/validation.py)!
const MAX_NAME_LENGTH = 16;
const NAME_LEGAL_SYMBOLS = "-_";

// NB: If updating these functions, ensure that the Django functions are also updated (in api/validation.py)!
function isValidName(name) {
	var valid = true;
	var reason = "";
	if (typeof name !== "string") {
		valid = false;
		reason = "Please enter a username.";
	} else if (name === "") {
		valid = false;
		reason = "Please enter a username.";
	} else if (name.length > MAX_NAME_LENGTH) {
		valid = false;
		reason = "The username entered is too long.";
	} else {
		for (var i = 0; i < name.length; i++)
			if (!isValidNameChar(name[i])) {
				valid = false;
				reason = "The username contains the ilvalid character '" + name[i] + "'."
				break;
			}
	}
	return {valid: valid, reason: reason};
}

// NB: If updating these functions, ensure that the Django functions are also updated (in api/validation.py)!
function isValidNameChar(ch) {
	var cd = function(s) { return s.charCodeAt(0); };
	var c = ch.charCodeAt(0);
	if (c >= cd('a') && c <= cd('z')) {
		// lowercase chars
		return true;
	} else if (c >= cd('A') && c <= cd('Z')) {
		// uppercase chars
		return true;
	} else if (c >= cd('1') && c <= cd('9')) {
		// digits
		return true;
	} else if (NAME_LEGAL_SYMBOLS.indexOf(ch) !== -1) {
		// symbols
		return true;
	}
	return false;
}
